File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.py
import heapq

def solution(operations):
    min_h = []  # (value, id)
    max_h = []  # (-value, id)
    alive = [False] * len(operations)

    next_id = 0
    valid = 0

    def clean_min():
        while min_h and not alive[min_h[0][1]]:
            heapq.heappop(min_h)

    def clean_max():
        while max_h and not alive[max_h[0][1]]:
            heapq.heappop(max_h)

    for op in operations:
        if op[0] == 'I':
            x = int(op[2:])
            idx = next_id
            next_id += 1
            alive[idx] = True
            heapq.heappush(min_h, (x, idx))
            heapq.heappush(max_h, (-x, idx))
            valid += 1
        else:  # 'D'
            if valid == 0:
                continue

            if op[2] == '1':  # delete max
                clean_max()
                if max_h:
                    _, idx = heapq.heappop(max_h)
                    alive[idx] = False
                    valid -= 1
            else:  # delete min (D -1)
                clean_min()
                if min_h:
                    _, idx = heapq.heappop(min_h)
                    alive[idx] = False
                    valid -= 1

    if valid == 0:
        return [0, 0]

    clean_min()
    clean_max()

    return [-max_h[0][0], min_h[0][0]]

File: 프로그래머스/3/42628. 이중우선순위큐/이중우선순위큐.kt
import java.util.PriorityQueue

class Solution {
    fun solution(operations: Array<String>): IntArray {
        // (value, id)
        val minH = PriorityQueue<Pair<Int, Int>>(compareBy { it.first })
        val maxH = PriorityQueue<Pair<Int, Int>>(compareByDescending { it.first })

        val alive = BooleanArray(operations.size)
        var nextId = 0
        var validCount = 0

        fun cleanMin() {
            while (minH.isNotEmpty() && !alive[minH.peek().second]) minH.poll()
        }
        fun cleanMax() {
            while (maxH.isNotEmpty() && !alive[maxH.peek().second]) maxH.poll()
        }

        for (op in operations) {
            if (op[0] == 'I') {
                val x = op.substring(2).toInt()
                val id = nextId++
                alive[id] = true
                minH.add(Pair(x, id))
                maxH.add(Pair(x, id))
                validCount++
            } else { // 'D'
                if (validCount == 0) continue

                if (op[2] == '1') { // delete max
                    cleanMax()
                    if (maxH.isNotEmpty()) {
                        val id = maxH.poll().second
                        alive[id] = false
                        validCount--
                    }
                } else { // delete min
                    cleanMin()
                    if (minH.isNotEmpty()) {
                        val id = minH.poll().second
                        alive[id] = false
                        validCount--
                    }
                }
            }
        }

        if (validCount == 0) return intArrayOf(0, 0)

        cleanMin()
        cleanMax()
        return intArrayOf(maxH.peek().first, minH.peek().first)
    }
}